### PR TITLE
VMware: new module: vmware_dvs_portgroup_facts

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py
test/integration/targets/vmware_dvs_portgroup_facts/aliases
test/integration/targets/vmware_dvs_portgroup_facts/tasks/main.yml

##### ANSIBLE VERSION

```
2.8-devel
```

### CODE REVIEW

**Comment by Akasurde** [comment]
@dericcrago @pdellaert Could you please review this PR ? I am segregating facts DVS portgroups on based DVS names, I would like to know if it looks good or flat facts about portgroups without DVS names looks good. Thanks.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    type: str
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- name: Gather DVS portgroup facts
+  vmware_dvs_portgroup_facts:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
```
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py]: You should add datacenter and register to the example.

**Comment by ckotte** [comment]
I can access specific elements via:
```
- name: vCenter | Gather DVS portgroup facts
  vmware_dvs_portgroup_facts:
    hostname: "{{ inventory_hostname }}"
    username: "{{ vcsa_username }}"
    password: "{{ vcsa_password }}"
    datacenter: "{{ item }}"
    validate_certs: "{{ validate_vcenter_certs }}"
  delegate_to: localhost
  register: dvs_portgroup_facts
  with_items: "{{ vcenter_datacenters }}"

- name: vCenter | DVS portgroups
  debug:
    msg: "{{ dvs_portgroup_facts.results[1].dvs_portgroup_facts['dvSwitch']['dvPortgroup'] }}"
```

I can loop through all portgroups of a dvSwitch and access specific properties via:
```
- name: vCenter | DVS portgroups
  debug:
    msg: "{{ item.value.port_policy.block_override }}"
  with_dict: "{{ dvs_portgroup_facts.results[1].dvs_portgroup_facts['dvSwitch'] }}"
```

Wouldn't it be easier to put the portgroup dicts in an array? Those examples should be documented in the examples section as well.

The module is quiet slow. Maybe it makes sense to add the option to disable policy gathering.

An option to filter for the DVS would also be great to speed it up in a big environment.

File: lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  delegate_to: localhost
+'''
```
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup_facts.py:72]: Could you add examples on how to access specific items or loop through portgroups?

**Comment by Akasurde** [comment]
@ckotte I will work on review comments.

**Comment by pdellaert** [comment]
No comments from my end. i support those that are there

**Comment by Akasurde** [comment]
@ckotte Could you please review this again ? Thanks.

**Comment by Akasurde** [comment]
@ckotte Any news ?

**Comment by Akasurde** [comment]
rebuild_merge

**Comment by dx0xm** [comment]
Hi all,

two items that I found very useful and needed to add:
1. wouldn't be great to add a variable for particular dvswitch?
2. wouldn't be great to get also the vlan information of the port group?

I added it to the code (Attached)

for item 1, adding new arg similar to datacenter and building a list (default is all to keep module as it was)

        if self.dvs_name == 'all':
            dvs_lists = get_all_objs(self.content, [vim.DistributedVirtualSwitch], folder=datacenter.networkFolder)
        else:
            dvsn = find_dvs_by_name(self.content, self.dvs_name)
            if dvsn is None:
                self.module.fail_json(msg="Failed to find the dvswitch %s" % self.dvs_name)
            else:
                dvs_lists=[dvsn]

for item 2, adding new entry to the existing construct without modifying much:

                if self.params['show_vlan_info']:
                    vlanInfo=dvs_pg.config.defaultPortConfig.vlan
                    if isinstance(vlanInfo,vim.dvs.VmwareDistributedVirtualSwitch.TrunkVlanSpec):
                        vlan_id_list = []
                        for vli in vlanInfo.vlanId:
                            vlan_id_list.append(str(vli.start)+"-"+str(vli.end))
                        vlan_info = dict(
                            trunk=True,
                            vlan_id=vlan_id_list
                        )
                    else:
                        vlan_info = dict(
                            trunk=False,
                            vlan_id=vlanInfo.vlanId
                        )


is working for me so hopefully will help someone else.


[vmware_dvs_portgroup_facts_option.zip](https://github.com/ansible/ansible/files/3375103/vmware_dvs_portgroup_facts_option.zip)

**Pull request merged**